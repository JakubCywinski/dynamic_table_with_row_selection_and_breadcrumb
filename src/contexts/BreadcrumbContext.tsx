import React, {
    createContext,
    useContext,
    useEffect,
    useState,
    ReactNode,
} from "react";

interface BreadcrumbContextValue {
    path: string[];
    setPath: (path: string[]) => void;
}

const BreadcrumbContext = createContext<BreadcrumbContextValue | undefined>(
    undefined
);

interface BreadcrumbProviderProps {
    children: ReactNode;
}

export const BreadcrumbProvider: React.FC<BreadcrumbProviderProps> = ({
    children,
}) => {
    const [path, setPath] = useState<string[]>(() => {
        const savedPath = localStorage.getItem("breadcrumbPath");
        return savedPath ? JSON.parse(savedPath) : [];
    });

    useEffect(() => {
        localStorage.setItem("breadcrumbPath", JSON.stringify(path));
    }, [path]);

    return (
        <BreadcrumbContext.Provider value={{ path, setPath }}>
            {children}
        </BreadcrumbContext.Provider>
    );
};

export const useBreadcrumb = (): BreadcrumbContextValue => {
    const context = useContext(BreadcrumbContext);
    if (!context) {
        throw new Error(
            "useBreadcrumb must be used within a BreadcrumbProvider"
        );
    }
    return context;
};
