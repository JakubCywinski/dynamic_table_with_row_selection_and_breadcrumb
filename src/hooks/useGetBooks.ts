import { useQuery, UseQueryOptions } from "@tanstack/react-query";
import { booksAPI } from "../services/api";

interface VolumeInfo {
    title: string;
    authors: string[];
    description: string;
    canonicalVolumeLink: string;
    imageLinks: {
        smallThumbnail: string;
    };
    // add other fields as needed
}

interface Book {
    id: string;
    volumeInfo: VolumeInfo;
}

interface BooksResponse {
    kind: string;
    items: Book[];
}

interface MultipleBooksResponse {
    [query: string]: BooksResponse;
}

const getBooks = async ({
    queryKey,
}: {
    queryKey: string[];
}): Promise<MultipleBooksResponse> => {
    const searchTerms = queryKey[0].split(",");
    const fetchPromises = searchTerms.map((term) => {
        const isAuthorQuery = term.startsWith("inauthor:");
        const params = isAuthorQuery ? { q: term } : { q: `subject:${term}` };
        return booksAPI.get<BooksResponse>("/volumes", { params });
    });
    const results = await Promise.all(fetchPromises);

    const response: MultipleBooksResponse = {};
    searchTerms.forEach((term, index) => {
        response[term] = results[index].data;
    });

    return response;
};

export const useGetBooks = (
    searchTerms: string[] = [],
    options?: UseQueryOptions<
        MultipleBooksResponse,
        Error,
        MultipleBooksResponse,
        [string]
    >
) => {
    const queryKey: [string] = [searchTerms.join(",")];
    return useQuery(
        queryKey,
        ({ queryKey: [keys] }) => getBooks({ queryKey: keys.split(",") }),
        options
    );
};
