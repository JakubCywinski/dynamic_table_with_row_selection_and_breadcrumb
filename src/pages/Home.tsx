import { Categories } from "@/components";

export const Home = () => {
    return (
        <section className="w-full h-[60vh] px-8 flex flex-col justify-center">
            <div className="text-center flex items-center flex-col">
                <h1 className="text-4xl font-bold max-w-xl">
                    Dynamic Table with Row Selection and Breadcrumb
                </h1>
                <p className=" mt-4">Try it now by choosing book category!</p>
            </div>
            <div>
                <Categories />
            </div>
        </section>
    );
};
