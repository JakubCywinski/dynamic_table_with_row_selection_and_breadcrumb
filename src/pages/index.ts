export { Home } from "./Home";
export { Books } from "./Books";
export { BooksDetails } from "./BooksDetails";
