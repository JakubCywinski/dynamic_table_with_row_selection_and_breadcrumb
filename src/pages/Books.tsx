import { FC } from "react";
import { useParams } from "react-router-dom";
import { useGetBooks } from "../hooks/useGetBooks";
import { BooksTable, Preloader } from "@/components";

type BookParams = {
    category: string;
};

export const Books: FC = () => {
    const { category } = useParams<BookParams>();

    if (!category) {
        return <div>Category not found</div>;
    }

    const { data, isLoading, isError } = useGetBooks([category]);

    if (isLoading) {
        return <Preloader />;
    }

    if (isError || !data) {
        return <div>Error occurred while fetching books.</div>;
    }

    const books = data[category]?.items || [];

    return <BooksTable books={books} />;
};
