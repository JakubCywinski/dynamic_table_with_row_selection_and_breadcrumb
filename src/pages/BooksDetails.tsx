import { FC } from "react";
import { useParams } from "react-router-dom";
import { useGetBooks } from "../hooks/useGetBooks";
import { BooksTable } from "@/components";
import { Preloader } from "@/components";

type BookParams = {
    author: string;
};

export const BooksDetails: FC = () => {
    const { author } = useParams<BookParams>();

    if (!author) {
        return <div>author not found</div>;
    }

    const { data, isLoading, isError } = useGetBooks([author]);

    if (isLoading) {
        return <Preloader />;
    }

    if (isError || !data) {
        return <div>Error occurred while fetching books.</div>;
    }

    const books = data[author]?.items || [];

    return <BooksTable books={books} />;
};
