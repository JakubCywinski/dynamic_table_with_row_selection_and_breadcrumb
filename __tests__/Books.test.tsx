import { render, screen } from "@/utils/utils";
import "@testing-library/jest-dom";
import { Books } from "@/pages";
import { it, describe, vi } from "vitest";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { BreadcrumbProvider } from "@/contexts/BreadcrumbContext";
import { useGetBooks } from "@/hooks/useGetBooks";

vi.mock("@/hooks/useGetBooks");

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            refetchOnWindowFocus: false,
            refetchOnReconnect: false,
            retry: false,
            staleTime: 5 * 60 * 1000,
        },
    },
});

describe("Books", () => {
    afterEach(() => {
        vi.clearAllMocks();
    });

    it("renders loading state", async () => {
        (useGetBooks as jest.Mock).mockReturnValue({ isLoading: true });

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={["/books/fiction"]}>
                    <BreadcrumbProvider>
                        <Routes>
                            <Route
                                path="/books/:category"
                                element={<Books />}
                            />
                        </Routes>
                    </BreadcrumbProvider>
                </MemoryRouter>
            </QueryClientProvider>
        );

        expect(screen.getByText(/loading.../i)).toBeInTheDocument();
    });

    it("renders error state", () => {
        (useGetBooks as jest.Mock).mockReturnValue({ isError: true });

        render(
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={["/books/fiction"]}>
                    <BreadcrumbProvider>
                        <Routes>
                            <Route
                                path="/books/:category"
                                element={<Books />}
                            />
                        </Routes>
                    </BreadcrumbProvider>
                </MemoryRouter>
            </QueryClientProvider>
        );

        expect(screen.getByText(/error/i)).toBeInTheDocument();
    });
});
