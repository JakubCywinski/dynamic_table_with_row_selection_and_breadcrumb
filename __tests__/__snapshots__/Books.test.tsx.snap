// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Books > renders loading state 1`] = `
<div>
  Category not found
</div>
`;

exports[`Books > renders match snapshot 1`] = `
<div>
  Category not found
</div>
`;
